api/envoy/type/matcher/v3/filter_state.proto

```diff
@@ -25,5 +27,8 @@ message FilterStateMatcher {
 
     // Matches the filter state object as a string value.
     StringMatcher string_match = 2;
+
+    // Matchers the filter state object as a ip Instance.
+    xds.core.v3.CidrRange ip_range = 3;
```

Comment:
All exist `CidrRange`s in the envoy api repo now is `config.core.v3.CidrRange`, please use it for now before we sort out a way to migrate all common protos to xds repo.

And consider to naming it address_match or something. And should `repeated` be used to allow configuring multiple ranges?

Reply:
It's not possible to use `config.core.v3.CidrRange` as it creates a circular dependency between `api/envoy/config/core/v3:pkg` and `api/envoy/type/matcher/v3:pkg`.

Reply:
> It's not possible to use config.core.v3.CidrRange as it creates a circular dependency between api/envoy/config/core/v3:pkg and api/envoy/type/matcher/v3:pkg

Then this is another problem we should resolve. Using xds one shouldn't be a way to resolve the circular dependency. Because we will encounter similar problem between the matcher pkg and core pkg in the xds repo.

Reply:
We're not going to move envoy-specific matchers to the xds repo, right? This matcher I'm adding would be envoy specific-- it's using filter state as a source. Thus I don't think there would be a risk of this specific circular dependency happening again. 

Do you have a alternative solution to address this? The solutions that have been discussed other than this seem to involve a disproportionately large amount of work compared to what I'm actually trying to achieve, which is just to add this matcher.

Reply:
> We're not going to move envoy-specific matchers to the xds repo, right? This matcher I'm adding would be envoy specific-- it's using filter state as a source. Thus I don't think there would be a risk of this specific circular dependency happening again. 
> 
> 
> 
> Do you have a alternative solution to address this? The solutions that have been discussed other than this seem to involve a disproportionately large amount of work compared to what I'm actually trying to achieve, which is just to add this matcher.

I am not mean this new field will bring problem to xds repo. I mean xds repo may have similar dependency problem in the future with we adding more and more API. And the way we used now to resolve the dependency problem is not the perfect/correct way.

Any change to the core API should be carefully considered. I will try to create a PR to use a more fine-grained pkg to resolve the dependency problem.

Reply:
I agree that we ultimately need to change the structure of the BUILD rules to be more granular, and that doing so would address this particular case.  However, I think that change is going to take a while to make, because there are a lot of external users that depend on the current build rules that will break when we change them.  I think this needs some careful thought to do it in a way that minimizes disruption for API users.

While we're considering the best way to do that, I think using `CidrRange` in the cncf/xds repo is a fine short-term solution.  The `CirdRange` proto already exists there, which means it can already be used, so we're not introducing any new problem here.  And in the long run, we ultimately want to move all of the proto files to that repo anyway.

Reply:
> While we're considering the best way to do that, I think using CidrRange in the cncf/xds repo is a fine short-term solution. The CirdRange proto already exists there, which means it can already be used, so we're not introducing any new problem here. And in the long run, we ultimately want to move all of the proto files to that repo anyway.

What I am worring is that `long run` is too long and we finally cannot achieve our target but leave envoy a bunch of chaos.

Because I cannot think of any way to migrade so much APIs to `xds.` and without effect Envoy. And also didn't see a related viable plan/roadmap. So, I will prefer the tranditional API until there is actual viability.

> I agree that we ultimately need to change the structure of the BUILD rules to be more granular, and that doing so would address this particular case. However, I think that change is going to take a while to make, because there are a lot of external users that depend on the current build rules that will break when we change them. I think this needs some careful thought to do it in a way that minimizes disruption for API users.

What a tragic... Yeah, at least we may change the pkg of golang.

Reply:
I definitely understand your concern about leaving envoy a mess. But for this change, for clarity, are you saying you would still prefer a different solution rather than using the xds api? Or do you think this is fine to use for this PR but elsewhere we should continue to use the envoy apis where possible?

Reply:
> I definitely understand your concern about leaving envoy a mess. But for this change, for clarity, are you saying you would still prefer a different solution rather than using the xds api? Or do you think this is fine to use for this PR but elsewhere we should continue to use the envoy apis where possible?

I think we have no choice except to compromise 😢 . Because as @markdroth said, after a quick check, there is no simple way to resolve the dependency problems because lots of third-parties depends on our current build rules. So, we must do it carefully with a full discussion.

So, to unblock this PR, let continue with `xds.` first.